### PR TITLE
prov/mrail: Add basic support for counters.

### PR DESCRIPTION
**Pull request by yburette**
FI_RECV, FI_SEND, FI_READ and FI_WRITE supported. An additional protocol would
be needed in order to support FI_REMOTE_READ and FI_REMOTE_WRITE.

Signed-off-by: yohann <yohann.burette@intel.com>

### CODE REVIEW

**Comment by shefty** [comment]
I would discourage trying to support counters with remote read/write.  That will result in a significant performance impact.

File: prov/mrail/src/mrail.h [hunk 1]
```diff
@@ -353,3 +353,17 @@ void mrail_free_req(struct mrail_ep *mrail_ep, struct mrail_req *req)
...
+               cntr->cntr_fid.ops->add(&cntr->cntr_fid, 1);
+       }
+}
+
+static inline void mrail_cntr_incerr(struct util_cntr *cntr)
+{
+       if (cntr) {
+               cntr->cntr_fid.ops->adderr(&cntr->cntr_fid, 1);
+       }
+}
```
Comment (shefty) [inline prov/mrail/src/mrail.h]: For success counters, I would use function pointers to increment functions, with a no-op version to handle the case where no counter has been assigned.  The error counter could be different and continue to have a check, so we can avoid needing function pointers for unexpected cases.

Comment (yburette) [reply prov/mrail/src/mrail.h]: I just pushed some changes to add the function pointers within the util_ep. Is that how you envisioned it?

Comment (shefty) [reply prov/mrail/src/mrail.h]: Yes, thanks

**Comment by shefty** [comment]
@aingerson @a-ilango @dmitrygladkov - please see if the generic changes work for the util providers.

File: prov/mrail/src/mrail_cq.c [hunk 2]
```diff
@@ -351,6 +353,11 @@ static void mrail_handle_rma_completion(struct util_cq *cq,
 			assert(0);
 		}
 
+		if (comp->flags & FI_TRANSMIT)
```
Comment (a-ilango) [inline prov/mrail/src/mrail_cq.c]: Shouldn't this be FI_WRITE?

File: prov/mrail/src/mrail_ep.c [hunk 3]
```diff
@@ -382,6 +382,9 @@ mrail_send_common(struct fid_ep *ep_fid, const struct iovec *iov, void **desc,
 		FI_WARN(&mrail_prov, FI_LOG_EP_DATA,
 			"Unable to fi_sendmsg on rail: %" PRIu32 "\n", i);
 		goto err2;
+	} else if (!(flags & FI_COMPLETION)) {
+		/* Inject won't generate a completion, so we count it here */
```
Comment (a-ilango) [inline prov/mrail/src/mrail_ep.c]: The comment can be dropped. Only absence of FI_COMPLETION would not generate completion. Passing only FI_INJECT flag can still generate completion.

Comment (yburette) [reply prov/mrail/src/mrail_ep.c]: Gotcha, I'll update the comment to say that no completion was requested. Does that work for you?

**Pull request merged**